packages/xsnap/src/xsnap.c

```diff
@@ -0,0 +1,1350 @@
+	}
+}
+
+#endif /* mxDebug */
+
+
+/**
+ * @param dest: where to store address of newly allocated buffer,
+ *		  which is null-terminated for compatibility with C strings
+ * @returns count of bytes read or -1 for error;
+ *			caller takes ownership of *dest unless an error occurs
+ */
+static int fxReadNetString(FILE *inStream, char** dest, size_t* len)
+{
+	int code = 0;
+	char* buf = NULL;
+
+	if (fscanf(inStream, "%9lu", len) < 1) {
+		/* >999999999 bytes is bad */
+		code = 1;
+	} else if (fgetc(inStream) != ':') {
+		code = 2;
+	} else {
+		buf = malloc(*len + 1); /* malloc(0) is not portable */
+		if (!buf) {
+			code = 3;
+		} else if (fread(buf, 1, *len, inStream) < *len) {
+			code = 4;
+		} else if (fgetc(inStream) != ',') {
+			code = 5;
+		} else {
+			*(buf + *len) = 0;
+		}
+		if (code != 0) {
+			free(buf);
+		}
+		*dest = buf;
+	}
+	return code;
+}
+
+static char* fxReadNetStringError(int code)
+{
+	switch (code) {
+	case 0: return "OK";
+	case 1: return "Cannot read netstring, reading length prefix, fscanf";
+	case 2: return "Cannot read netstring, invalid delimiter or end of file, fgetc";
+	case 3: return "Cannot read netstring, cannot allocate message buffer, malloc";
+	case 4: return "Cannot read netstring, cannot read message body, fread";
+	case 5: return "Cannot read netstring, cannot read trailer, fgetc";
+	default: return "Cannot read netstring";
+	}
+}
+
+static int fxWriteNetString(FILE* outStream, char prefix, char* buf, size_t length)
+{
+	if (fprintf(outStream, "%lu:%c", length + 1, prefix) < 1) {
+		return 1;
+	} else if (fwrite(buf, 1, length, outStream) < length) {
+		return 2;
+	} else if (fputc(',', outStream) == EOF) {
+		return 3;
+	} else if (fflush(outStream) < 0) {
+		return 4;
+	}
+
+	return 0;
+}
+
+static char* fxWriteNetStringError(int code)
+{
+	switch (code) {
+	case 0: return "OK";
+	case 1: return "Cannot write netstring, error writing length prefix";
+	case 2: return "Cannot write netstring, error writing message body";
+	case 3: return "Cannot write netstring, error writing terminator";
+	case 4: return "Cannot write netstring, error flushing stream, fflush";
+	default: return "Cannot write netstring";
+	}
+}
+
+/**
+ * sysCall: write netstring; read netstring reply
+ */
+static void fx_sysCall(xsMachine *the)
+{
+	int argc = xsToInteger(xsArgc);
+	if (argc < 1) {
+		mxTypeError("expected ArrayBuffer");
+	}
+
+	size_t length;
+	char* buf = NULL;
+	txSlot* arrayBuffer = &xsArg(0);
+	length = fxGetArrayBufferLength(the, arrayBuffer);
+
+	buf = malloc(length);
+
+	fxGetArrayBufferData(the, arrayBuffer, 0, buf, length);
+	int writeError = fxWriteNetString(toParent, '?', buf, length);
+
+	free(buf);
+
+	if (writeError != 0) {
+		xsUnknownError(fxWriteNetStringError(writeError));
+	}
+
+	// read netstring
+	size_t len;
+	int readError = fxReadNetString(fromParent, &buf, &len);
+	if (readError != 0) {
+		xsUnknownError(fxReadNetStringError(readError));
+	}
+
+	xsResult = xsArrayBuffer(buf, len);
+}
```

Comment:
I refactored this to share the underlying fxReadNetString and fxWriteNetString.

Observing we always prepend a command code in the body of a netstring, I generalized fxWriteNetString to always include the command code and variable length buffer, to avoid allocating an intermediate buffer one byte longer in some cases.

Reply:
wait... the `xsnap.c` in 60d36908f1 is not a copy of `xsnap.c` before we started messing with it? I would much prefer that our changes were in a separate commit. But I guess it's not critical.

Reply:
Introduced a commit with the state of each file before we touched them so our changes show in relief.